### PR TITLE
Init Resource Handlers for AWS::CodeArtifact::Repository

### PR DESCRIPTION
**Pull request by jonjara**
Basic implementation of Create/Read/Delete handlers.

**Description of changes:**
Template code from `cfn-init` and implemented Create/Delete/Read handlers. 

**Testing**
Did `cfn-submit` on my own account, created a stack with a repository/domain template and deleted it.

Some TODOs:
+ update/list handlers
+ Add upstream, external connection, and policy handling for create/update handlers and to the resource schema

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-codeartifact-repository/aws-codeartifact-repository.json [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  ],
+  "primaryIdentifier": [
+    "/properties/Arn"
+  ],
+  "handlers": {
+    "create": {
+      "permissions": [
+        "codeartifact:CreateRepository",
+        "codeartifact:DescribeRepository"
+      ]
```
Comment (joeymarylander) [inline aws-codeartifact-repository/aws-codeartifact-repository.json:60]: We also want `codeartifact:PutRepositoryPermissionPolicy`

File: aws-codeartifact-repository/aws-codeartifact-repository.json [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    "read": {
+      "permissions": [
+        "codeartifact:DescribeRepository"
+      ]
+    },
+    "update": {
+      "permissions": [
+        "codeartifact:PutRepositoryPermissionsPolicy",
+        "codeartifact:DeleteRepositoryPermissionsPolicy"
+      ]
```
Comment (joeymarylander) [inline aws-codeartifact-repository/aws-codeartifact-repository.json:71]: Also we want `codeartifact:UpdateRepository`

File: aws-codeartifact-repository/resource-role.yaml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            Version: '2012-10-17'
+            Statement:
+              - Effect: Allow
+                Action:
+                - "codeartifact:CreateRepository"
+                - "codeartifact:DeleteRepository"
+                - "codeartifact:DeleteRepositoryPermissionsPolicy"
+                - "codeartifact:DescribeRepository"
+                - "codeartifact:ListRepositories"
+                - "codeartifact:PutRepositoryPermissionsPolicy"
```
Comment (joeymarylander) [inline aws-codeartifact-repository/resource-role.yaml:31]: also updaterepo

File: aws-codeartifact-repository/src/main/java/software/amazon/codeartifact/repository/ReadHandler.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        return proxy.initiate("AWS-CodeArtifact-Domain::Read", proxyClient, request.getDesiredResourceState(), callbackContext)
+            // STEP 2 [construct a body of a request]
+            .translateToServiceRequest(Translator::translateToReadRequest)
+            // STEP 3 [make an api call]
+            .makeServiceCall((awsRequest, client) -> {
+                DescribeRepositoryResponse awsResponse = null;
+                try {
+                    awsResponse = client.injectCredentialsAndInvokeV2(awsRequest, proxyClient.client()::describeRepository);
+                } catch (final AwsServiceException e) {
+                    String arn = request.getDesiredResourceState().getRepositoryName();
```
Comment (joeymarylander) [inline aws-codeartifact-repository/src/main/java/software/amazon/codeartifact/repository/ReadHandler.java:34]: thats not an arn

File: aws-codeartifact-repository/aws-codeartifact-repository.json [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    "AdministratorAccount": {
+      "description": "The 12-digit account ID of the AWS account that manages the repository.",
+      "pattern": "[0-9]{12}",
+      "type": "string"
+    },
+    "Description": {
+      "description": "A text description of the repository.",
+      "type": "string",
+      "maxLength": 100
+    },
```
Comment (joeymarylander) [inline aws-codeartifact-repository/aws-codeartifact-repository.json:34]: This isn't right. A repo description can be much longer than this, please update based on our model.

File: aws-codeartifact-repository/aws-codeartifact-repository.json [hunk 6]
```diff
@@ -0,0 +1,84 @@
+{
```
Comment (joeymarylander) [inline aws-codeartifact-repository/aws-codeartifact-repository.json:1]: please add the appropriate `required` properties.

File: aws-codeartifact-repository/aws-codeartifact-repository.json (hunk 5 above)
Comment (jonjara) [reply aws-codeartifact-repository/aws-codeartifact-repository.json:34]: missed a 0, thanks for catching

File: aws-codeartifact-repository/src/main/java/software/amazon/codeartifact/repository/ReadHandler.java (hunk 4 above)
Comment (jonjara) [reply aws-codeartifact-repository/src/main/java/software/amazon/codeartifact/repository/ReadHandler.java:34]: true, will update

File: aws-codeartifact-repository/aws-codeartifact-repository.json (hunk 6 above)
Comment (jonjara) [reply aws-codeartifact-repository/aws-codeartifact-repository.json:1]: will do

File: aws-codeartifact-repository/.rpdk-config [hunk 7]
```diff
@@ -0,0 +1,17 @@
+{
+    "typeName": "AWS::CodeArtifact::Repository",
+    "language": "java",
+    "runtime": "java8",
```
Comment (detam-amazon) [inline aws-codeartifact-repository/.rpdk-config:4]: Are we updating this to a newer version of Java?

File: aws-codeartifact-repository/aws-codeartifact-repository.json [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  "typeName": "AWS::CodeArtifact::Repository",
+  "description": "The resource schema to create a CodeArtifact repository.",
+  "sourceUrl": "https://github.com/aws-cloudformation/aws-cloudformation-resource-providers-codeartifact",
+  "properties": {
+    "RepositoryName": {
+      "description": "The name of the repository.",
+      "type": "string",
+      "pattern": "[a-z][a-z0-9\\-]{0,48}[a-z0-9]",
```
Comment (detam-amazon) [inline aws-codeartifact-repository/aws-codeartifact-repository.json:9]: I believe the pattern for repo name is different:
https://alpha-docs-aws.amazon.com/codeartifact/latest/APIReference/API_CreateRepository.html

File: aws-codeartifact-repository/src/main/java/software/amazon/codeartifact/repository/DeleteHandler.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+                    .success());
+    }
+
+    private AwsResponse deleteRepository(
+        ProgressEvent<ResourceModel, CallbackContext>  progress,
+        ProxyClient<CodeartifactClient> client,
+        DeleteRepositoryRequest awsRequest
+    ) {
+        AwsResponse awsResponse = null;
+        String arn = progress.getResourceModel().getRepositoryName();
```
Comment (detam-amazon) [inline aws-codeartifact-repository/src/main/java/software/amazon/codeartifact/repository/DeleteHandler.java:45]: Does getRepositoryName return arn? or is the variable misnamed?
Also, you probably only need to make this call in the catch.

Comment (jonjara) [reply aws-codeartifact-repository/src/main/java/software/amazon/codeartifact/repository/DeleteHandler.java:45]: misnamed, I'll update.

File: aws-codeartifact-repository/aws-codeartifact-repository.json (hunk 8 above)
Comment (jonjara) [reply aws-codeartifact-repository/aws-codeartifact-repository.json:9]: oh yeah it is, thanks

File: aws-codeartifact-repository/.rpdk-config (hunk 7 above)
Comment (jonjara) [reply aws-codeartifact-repository/.rpdk-config:4]: I'll create an issue to track this

**Pull request merged**